12-multiplatform/3_tinycity/3_3_pedestrian_button/main.go

```diff
@@ -0,0 +1,137 @@
+package main
+
+import (
+	"time"
+
+	"machine"
+)
+
+func main() {
+	car := newCarLight(machine.D2, machine.D3, machine.D4)
+	walk := newWalkLight(machine.D5, machine.D6)
+	button := machine.D7
+	c := newCrossing(car, walk, button)
+
+	go c.listenButton()
+
+	for {
+		c.Switch()
+		select {
+		case <-c.buttonPressed:
+		case <-time.After(time.Second * 5):
+		}
+	}
+}
+
+type crossing struct {
+	cars          *carLight
+	walks         *walkLight
+	button        *machine.Pin
+	pedestriansGo bool
+	buttonPressed chan struct{}
+}
+
+func newCrossing(cars *carLight, walks *walkLight, button machine.Pin) *crossing {
+	button.Configure(machine.PinConfig{Mode: machine.PinInputPullup})
+
+	return &crossing{
+		cars:          cars,
+		walks:         walks,
+		button:        &button,
+		buttonPressed: make(chan struct{}, 1),
+	}
+}
+
+func (c *crossing) Switch() {
+	if c.pedestriansGo {
+		c.pedestriansGo = false
+		c.walks.Stop()
+		c.cars.Go()
+	} else {
+		c.pedestriansGo = true
+		c.cars.Stop()
+		c.walks.Go()
+	}
+}
+
+func (c *crossing) listenButton() {
+	for {
+		if !c.pedestriansGo && !c.button.Get() {
```

Comment:
I don't recall whether button behavior in tinygo is sticky or not. The reason it maters:

1. If it's not sticky -- if the button click lasts less than 100ms, then the button won't trigger. And clicking a button less than 100ms is within human limit.
2. if it is sticky -- then you need to worry about multiple clicks, i.e. it needs to be reset only after the walk cycle finishes.

Usually, button interrupts do work better, but probably would be too advanced for this book.

Reply:
In the [TinyGo playground code](https://github.com/tinygo-org/playground/blob/57def271b3ff47b8d2907814a26b641bd2b34329/worker/parts.js#L320), it states 
``// Button implements a typical push button (single pole, normally open).``

It should be sticky but I do not understand how the code is working since `!c.button.Get()` should return true when it is pressed...

I have tested it on my hardware and it does not work as is, I had to remove the `!` which makes more sense to me.

Reply:
https://tinygo.org/docs/reference/machine/ metnions that `Get` retrieves the momentary value of the switch. So it's not sticky. In that case I would recommend decreasing the polling rate below 10ms and maybe demonstrating the problem when you poll at 1s or 2s rate.